packages/byods/src/BYODS.ts

```diff
@@ -1,47 +1,389 @@
+}
+
+/**
+ * Represents a token with its expiration details.
+ * @interface
+ * @property {string} accessToken - The access token.
+ * @property {string} refreshToken - The refresh token.
+ * @property {Date} expiresAt - The expiration date of the access token.
+ * @property {Date} refreshAccessTokenExpiresAt - The expiration date of the refresh token.
+ */
+interface Token {
   accessToken: string;
   refreshToken: string;
   expiresAt: Date;
+  refreshAccessTokenExpiresAt: Date;
 }
 
-class BYODS {
-  private clientId: string;
-  private clientSecret: string;
-  private tokenHost: string;
-  private accessToken: string;
-  private refreshToken: string;
-  private expiresAt: Date;
+/**
+ * Represents a service app authorization for an organization.
+ * @interface
+ * @property {string} orgId - The organization ID.
+ * @property {Token} token - The token details.
+ */
+interface ServiceAppAuthorization {
+  orgId: string;
+  token: Token;
+}
+
+/**
+ * Represents a collection of service app authorizations.
+ * @interface
+ * @property {string} [orgId: string] - The organization ID mapped to its authorization details.
+ */
+interface ServiceAppAuthorizations {
+  [orgId: string]: ServiceAppAuthorization;
+}
+
+export default class BYODS {
+  private serviceAppId: string;
+  private baseUrl: string;
+  private headers: Record<string, string> = {
+    'User-Agent': userAgent,
+  };
+
+  private jwksCache: Record<string, never> = {};
+  private jwks: any;
+
+  private config: SDKConfig;
+  private serviceAppAuthorizations: ServiceAppAuthorizations = {};
+
+  constructor(clientId: string, clientSecret: string, baseUrl = defaultBaseUrl) {
+    this.baseUrl = baseUrl;
+    this.config = {clientId, clientSecret, baseUrl};
+    this.serviceAppId = Buffer.from(`ciscospark://us/APPLICATION/${clientId}`).toString('base64');
+
+    // Create a remote JWK Set
+    this.jwks = createRemoteJWKSet(new URL(jwksUrl), {
+      [jwksCache]: this.jwksCache,
+      cacheMaxAge: 600000, // 10 minutes
+      cooldownDuration: 30000, // 30 seconds
+    });
+  }
+
+  /**
+   * Save the service app registration using the provided refresh token.
+   * @param {string} orgId - The organization ID.
+   * @param {string} refreshToken - The refresh token.
+   * @returns {Promise<void>}
+   */
+  public async saveServiceAppRegistration(orgId: string, refreshToken: string): Promise<void> {
+    try {
+      const response = await fetch(`${this.baseUrl}/access_token`, {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/x-www-form-urlencoded',
+          ...this.headers,
+        },
+        body: new URLSearchParams({
+          grant_type: 'refresh_token',
+          client_id: this.config.clientId,
+          client_secret: this.config.clientSecret,
+          refresh_token: refreshToken,
+        }),
+      });
+
+      if (!response.ok) {
+        throw new Error(`Failed to save service app registration: ${response.statusText}`);
+      }
```

Comment:
Should this come in the catch block?

Reply:
This is consistent with what we are doing in calling. This is fine as we don't have to catch the error here, we already have received the failed response